python/thirdweb-ai/src/thirdweb_ai/services/storage.py

```diff
@@ -0,0 +1,176 @@
+import asyncio
+import hashlib
+import json
+import mimetypes
+import os
+from collections.abc import AsyncGenerator
+from dataclasses import asdict, is_dataclass
+from io import BytesIO
+from pathlib import Path
+from typing import Annotated, Any
+
+import httpx
+from pydantic import BaseModel
+
+from thirdweb_ai.services.service import Service
+from thirdweb_ai.tools.tool import tool
+
+
+async def async_read_file_chunks(file_path: str | Path, chunk_size: int = 8192) -> AsyncGenerator[bytes, None]:
+    """Read file in chunks asynchronously to avoid loading entire file into memory."""
+    async with asyncio.Lock():
+        path_obj = Path(file_path) if isinstance(file_path, str) else file_path
+        with path_obj.open("rb") as f:
+            while chunk := f.read(chunk_size):
+                yield chunk
+
+
+class Storage(Service):
+    def __init__(self, secret_key: str):
+        super().__init__(base_url="https://storage.thirdweb.com/ipfs", secret_key=secret_key)
```

Comment:
nit: the base_url should be `https://storage.thirdweb.com` and the endpoints should prepend `/ipfs`